### DISCUSSION TITLE
cache.resolve fails to resolve nodes inside relayPagination, fetching is true without end

### DISCUSSION THREAD

**Discussion by baransu** [urql-graphql/urql#2217] [Q&A] [answered]
We're running into a really weird issue that we failed to reproduce. Our `cacheExchange` gets into this broken state.

1. relay pagination query is performed with following arguments 
```js
 requestPolicy: "cache-and-network",
 pause: !teamId,
 variables: {
    "filters": {
        "assignedTo": [],
        "createdBy": [],
        "devices": [],
        "locations": [],
        "mentions": [],
        "projectIds": [],
        "statuses": [
            "Done",
            "ToDo"
        ]
    },
    "first": 50,
   // this is to make TS happy when using `pause`
    "teamId": teamId ?? "",
    "textSearch": textSearch
}
```
2. network request is never executed
3. in local resolvers `relayPagination` is set with default values
4. initial query is fired returning `{ fetching: true, stale: false, data: { edges: Array(5)} }`
5. when we investigate further `relayPagination` internally returns the correct number of 50 edges (as defined by query variables) 
6. when `relayPagination` fires `cache.resolve(edge, 'node') for each of those 50 edges it can resolve just 5 nodes
7. when we perform search which updates `textSearch` argument this solves this 'stale' problem

We have a few questions that we really don't understand about this situation:
1. Why query ends up in an infinite `fetching` state?
2. Why it can resolve only 5 out of 50 edges?
3. We use `pause` which some people reported is not working 100% correctly. May it be related?
4. We tried triggering `executeQuery` in `useEffect` to solve that in a hacky way but without luck - the network request never fired

I'm attaching `edges` before and after `cache.resolve`: https://gist.github.com/Faliszek/39acbc7ee4e8c6c1e23eefb62bb822c1 ("Resolved nodes in concatEdges" contains only 15 entries but all 50 are returned with just 5 nodes resolved correctly)

cc @Faliszek

**Comment by kitten** [comment, marked as answer]
It looks like in the sample data not all nodes are keyable? This is a little odd, since they should at least be complete pieces of data, so I'm not quite sure how to piece this together. It looks a bit like the `Edge.node` field isn't present... which seems... odd, given that yirue likely querying it?

We may need just a tad more information here, but do you have any warnings in your console from Graphcache that could help us here?

This may also explain the infinite fetching. It's almost as if some data is missing and Graphcache is infinitely trying to get it, again and again, which would very well explain this behaviour

**Reply by baransu** [reply]
Yes, the node is queried.

We haven't found any errors/warnings in the console from Graphcache or anything related to `urql`.

Here is exact query we're executing: https://gist.github.com/baransu/e3f7ff1b8742961d9429d5dfef002414

As I'm looking at that query I wonder if `id: uid` alias could be one to blame. Could it be that some part of Graphcache is ignoring that alias?

**Reply by kitten** [reply]
Have you specified a custom key field for this? i.e. `keys: { Thread: data => data.uid }`? It's worth noting that an alias is not a _real field_, but rather a _virtual field_, meaning, it doesn't change what the field is stored as or named internally.

The reason for that is the normalisation in general, while we may query `any_field_name: any_field`, in the next query it could then be `yet_another_field_name: any_field`, which both point to the same underlying field.

**Reply by baransu** [reply]
We used an alias as it solved warnings for us. We don't have a custom key field for it. 

So if I understand correctly alias could lead to problems with cache?

While I understand it's _unstable_ as it may change independently from the schema in our case we have it like this from the begging of the query, and also on all of the fragments/queries on the given `Thread` type, no one changed anything yet the problem still may occur.
